### PR TITLE
vSphere advanced capabilities with datastore support and api improvements

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
## Description
Added support for different datastore types like vSAN, vVols, VMFS6, Datastore clusters. Along with these native APIs of vCenter have been used.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@harikrishna-patnala most of the tests are failing for travis/simulator hinting at general failures - can you have a look at simulator failures and manually test vm lifecycle before pushing trillian to work (which takes hours)?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
Sure @rhtyd I'll check simulator failures

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@harikrishna-patnala when deploying zone (with simulator) this fails with:
```
Exception Occurred: ['Traceback (most recent call last):\n', '  File "tools/marvin/marvin/deployDataCenter.py", line 280, in createPrimaryStorages\n    ret = self.__apiClient.createStoragePool(primarycmd)\n', '  File "/usr/local/lib/python2.7/dist-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 3364, in createStoragePool\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/local/lib/python2.7/dist-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: createstoragepool failed, due to: errorCode: 530, errorText:Failed to add data store: Failed to access storage pool\n']





2:02
2020-08-07 14:01:28,340 WARN  [o.a.c.s.d.l.CloudStackPrimaryDataStoreLifeCycleImpl] (ApiServer-8:ctx-c2e3faa6 ctx-b78205c4) (logid:163a78d3) Unable to establish a connection between Host[-2-Routing] and org.apache.cloudstack.storage.datastore.PrimaryDataStoreImpl@19656adf
java.lang.NullPointerException
	at org.apache.cloudstack.storage.datastore.provider.DefaultHostListener.hostConnect(DefaultHostListener.java:106)
	at com.cloud.storage.StorageManagerImpl.connectHostToSharedPool(StorageManagerImpl.java:1027)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUs
```

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
re-kicked centos7 pkg manually

**Comment by harikrishna-patnala** [comment]
> re-kicked centos7 pkg manually

Thanks @rhtyd

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 1]
```diff
@@ -616,6 +616,18 @@ public void prepareIsoForVmProfile(VirtualMachineProfile profile, DeployDestinat
...
+            s_logger.debug("Schedule to preload template " + template.getId() + " into child datastores of DataStore cluster: " + pool.getId());
+            for (StoragePoolVO childDataStore :  childDataStores) {
+                prepareTemplateInOneStoragePoolInternal(template, childDataStore);
+            }
+        } else {
+            prepareTemplateInOneStoragePoolInternal(template, pool);
+        }
+    }
+
+    private void prepareTemplateInOneStoragePoolInternal(final VMTemplateVO template, final StoragePoolVO pool) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:630]: naming is kind of off; `private prepareTemplateInOneStoragePool(..)` is already internal as the fact that it is `private` suggests. I would think this name would point to what it actually does, like `startAThreadToPreloadTemplate(..)` or such a thing.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -2192,6 +2193,13 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
             throw new InvalidParameterValueException("Cannot migrate volume " + vol + "to the destination storage pool " + destPool.getName() + " as the storage pool is in maintenance mode.");
         }
 
+        if (destPool.getPoolType() == Storage.StoragePoolType.DatastoreCluster) {
+            DataCenter dc = _entityMgr.findById(DataCenter.class, vol.getDataCenterId());
+            Pod destPoolPod = _entityMgr.findById(Pod.class, destPool.getPodId());
+
+            destPool = _volumeMgr.findChildDataStoreInDataStoreCluster(dc, destPoolPod, destPool.getClusterId(), null, null, destPool.getId());
+        }
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2202]: `conditionallyTraverseDatastoreClusterToReachBasicDatastore(..)`

File: server/src/main/java/com/cloud/storage/StoragePoolAutomationImpl.java [hunk 3]
```diff
@@ -117,9 +117,11 @@ public boolean maintain(DataStore store) {
...
             }
             for (StoragePoolVO sp : spes) {
-                if (sp.getStatus() == StoragePoolStatus.PrepareForMaintenance) {
-                    throw new CloudRuntimeException("Only one storage pool in a cluster can be in PrepareForMaintenance mode, " + sp.getId() +
-                        " is already in  PrepareForMaintenance mode ");
+                if (sp.getParent() != pool.getParent()) { // If Datastore cluster is tried to prepare for maintenance then child storage pools are also kept in PrepareForMaintenance mode
+                    if (sp.getStatus() == StoragePoolStatus.PrepareForMaintenance) {
+                        throw new CloudRuntimeException("Only one storage pool in a cluster can be in PrepareForMaintenance mode, " + sp.getId() +
+                                " is already in  PrepareForMaintenance mode ");
+                    }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StoragePoolAutomationImpl.java:124]: good comment but it should be reflected in a method name, modularising the code further.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -3008,6 +3034,12 @@ public DiskOffering createDiskOffering(final CreateDiskOfferingCmd cmd) {
             }
         }
 
+        if (storagePolicyId != null) {
+            if (vsphereStoragePolicyDao.findById(storagePolicyId) == null) {
+                throw new InvalidParameterValueException("Please specify a valid vSphere storage policy id");
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:3041]: why double nesting? this is actually only one error condition is it?

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java [hunk 5]
```diff
@@ -1268,6 +1269,18 @@ public int compare(Volume v1, Volume v2) {
...
+                                try {
+                                    boolean isStoragePoolStoragepolicyComplaince = _storageMgr.isStoragePoolComplaintWithStoragePolicy(requestVolumes, potentialSPool);
+                                    if (!isStoragePoolStoragepolicyComplaince) {
+                                        continue;
+                                    }
+                                } catch (StorageUnavailableException e) {
+                                    s_logger.warn(String.format("Could not verify storage policy complaince against storage pool %s due to exception %s", potentialSPool.getUuid(), e.getMessage()));
+                                    continue;
+                                }
+                            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:1282]: `handleVmwareStoragePoolCompliance(..)`

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 6]
```diff
@@ -901,10 +925,54 @@ public boolean deletePool(DeletePoolCmd cmd) {
             s_logger.warn("Unable to delete storage id: " + id + " due to it is not in Maintenance state");
             throw new InvalidParameterValueException("Unable to delete storage due to it is not in Maintenance state, id: " + id);
         }
-        Pair<Long, Long> vlms = _volsDao.getCountAndTotalByPool(id);
+
+        if (sPool.getPoolType() == StoragePoolType.DatastoreCluster) {
+            // FR41 yet to handle on failure of deletion of any of the child storage pool
+            if (checkIfDataStoreClusterCanbeDeleted(sPool, forced)) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:931]: are we sure the check outside the transaction is secure/safe? even if it seems to be just the return, still in the check and during the stack drop context switches may have occurred. please reconsider.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 7]
```diff
@@ -1469,11 +1537,55 @@ public PrimaryDataStoreInfo preparePrimaryStorageForMaintenance(Long primaryStor
...
+
+        if (primaryStorage.getPoolType() == StoragePoolType.DatastoreCluster) {
+            handlePrepareDatastoreCluserMaintenance(lifeCycle, primaryStorageId);
+        }
         lifeCycle.maintain(store);
 
         return (PrimaryDataStoreInfo)_dataStoreMgr.getDataStore(primaryStorage.getId(), DataStoreRole.Primary);
     }
 
+    private void handlePrepareDatastoreCluserMaintenance(DataStoreLifeCycle lifeCycle, Long primaryStorageId) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1549]: rather large, but it be further split?

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 8]
```diff
@@ -1469,11 +1537,55 @@ public PrimaryDataStoreInfo preparePrimaryStorageForMaintenance(Long primaryStor
...
+        for (Iterator<StoragePoolVO> iteratorChildDatastore = childDatastores.listIterator(); iteratorChildDatastore.hasNext(); ) {
+            DataStore childStore = _dataStoreMgr.getDataStore(iteratorChildDatastore.next().getId(), DataStoreRole.Primary);
+            try {
+                lifeCycle.maintain(childStore);
+            } catch (Exception e) {
+                if (s_logger.isDebugEnabled()) {
+                    s_logger.debug(String.format("Exception on maintenance preparation of one of the child datastores in datastore cluster %d with error %s", primaryStorageId, e));
+                    s_logger.debug(String.format("Cancelling the maintenance mode of child datastores in datastore cluster %d", primaryStorageId));
+                }
+                // Cancel maintenance mode of already prepared child storage pools
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1572]: comments lie this are a tell that a method could be factored out.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 9]
```diff
@@ -1469,11 +1537,55 @@ public PrimaryDataStoreInfo preparePrimaryStorageForMaintenance(Long primaryStor
...
+                maintenanceSuccessfulStoragePools.add(childStore);
+                for (DataStore dataStore: maintenanceSuccessfulStoragePools) {
+                    lifeCycle.cancelMaintain(dataStore);
+                }
+                // Set back to Up state of remaining child storage pools
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1577]: another method name in comment ;)

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 10]
```diff
@@ -1496,6 +1608,14 @@ public PrimaryDataStoreInfo cancelPrimaryStorageForMaintenance(CancelPrimaryStor
         DataStoreProvider provider = _dataStoreProviderMgr.getDataStoreProvider(primaryStorage.getStorageProviderName());
         DataStoreLifeCycle lifeCycle = provider.getDataStoreLifeCycle();
         DataStore store = _dataStoreMgr.getDataStore(primaryStorage.getId(), DataStoreRole.Primary);
+        if (primaryStorage.getPoolType() == StoragePoolType.DatastoreCluster) {
+            //FR41 need to handle when one of the primary stores is unable to cancel the maintenance mode
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1612]: internal project code in comment, method extract needed.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 11]
```diff
@@ -1870,6 +1990,57 @@ public boolean storagePoolHasEnoughSpaceForResize(StoragePool pool, long current
         }
     }
 
+    @Override
+    public boolean isStoragePoolComplaintWithStoragePolicy(List<Volume> volumes, StoragePool pool) throws StorageUnavailableException {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1994]: 50 line, two for loops and quadruple nested blocks. please dissect.

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@harikrishna-patnala can you fix conflicts? thanks

**Comment by harikrishna-patnala** [comment]
> @harikrishna-patnala can you fix conflicts? thanks

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
Re-kicked manually

**Comment by rhtyd** [comment]
Ignore last test results, new env test on-going (job/acs-pr-trillian/2437/console)

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@harikrishna-patnala can you look at travis failures after the next run? Most of the jobs had failed.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
(lab down so BO etc will have to wait)

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@harikrishna-patnala can you look at build/Travis failure?

**Comment by harikrishna-patnala** [comment]
> @harikrishna-patnala can you look at build/Travis failure?

Any specific reason why this is failing on Travis, although it was successful on my local environment.

[ERROR] Failed to execute goal on project cloud-plugin-storage-volume-default: Could not resolve dependencies for project org.apache.cloudstack:cloud-plugin-storage-volume-default:jar:4.15.0.0-SNAPSHOT: Could not find artifact org.apache.cloudstack:cloud-plugin-hypervisor-vmware:jar:4.15.0.0-SNAPSHOT in apache.snapshots (http://repository.apache.org/snapshots) -> [Help 1]

This happened due to these changes
diff --git a/plugins/storage/volume/default/pom.xml b/plugins/storage/volume/default/pom.xml
index 51450e0f8a..c620e72ad3 100644
--- a/plugins/storage/volume/default/pom.xml
+++ b/plugins/storage/volume/default/pom.xml
@@ -33,6 +33,12 @@
             <artifactId>cloud-engine-storage-volume</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.cloudstack</groupId>
+            <artifactId>cloud-plugin-hypervisor-vmware</artifactId>
+            <version>4.15.0.0-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
     </dependencies>
     <build>

**Comment by rhtyd** [comment]
In that case let's rekick it

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
kicked a xenserver-71 env test manually, plus kicking a regression test run for vmware67u3
@blueorangutan test centos7 vmware-67u3

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test centos8 vmware-67u3

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test centos8 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@harikrishna-patnala when you've time can you check the trillian failure, are those env related or caused by any regression in this PR? For the Travis failure, if we move the VMware validation via a Command-Answer pattern that will fix the issue of noredist dependency (Travis does not build all jobs with -Dnoredist)

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@rhtyd, jenkins is failing with a git config error. can you please check that.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@harikrishna-patnala can you review and close this PR and continue work over https://github.com/apache/cloudstack/pull/4307 thnx

**Comment by harikrishna-patnala** [comment]
Closing this PR as this is merged with another feature on PR #4307

**Pull request closed without merging**